Content.Server/Body/Surgery/SurgerySystem.cs

```diff
@@ -0,0 +1,560 @@
+using System.Threading.Tasks;
+using Content.Server.DoAfter;
+using Content.Server.Notification;
+using Content.Shared.Body;
+using Content.Shared.Body.Part;
+using Content.Shared.Body.Surgery;
+using Content.Shared.Body.Surgery.Operation;
+using Content.Shared.Body.Surgery.Operation.Step;
+using Content.Shared.Body.Surgery.Surgeon;
+using Content.Shared.Body.Surgery.Target;
+using Content.Shared.Body.Surgery.Tool;
+using Content.Shared.Body.Surgery.UI;
+using Content.Shared.Interaction;
+using Content.Shared.Notification.Managers;
+using Robust.Server.GameObjects;
+using Robust.Server.Player;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Prototypes;
+
+namespace Content.Server.Body.Surgery
+{
+    public class SurgerySystem : SharedSurgerySystem
+    {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly ILocalizationManager _loc = default!;
+        [Dependency] private readonly IPlayerManager _playerManager = default!;
+
+        [Dependency] private readonly DoAfterSystem _doAfter = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<SurgeryDrapesComponent, ComponentStartup>(OnDrapesStartup);
+            SubscribeLocalEvent<SurgeryDrapesComponent, AfterInteractEvent>(OnDrapesAfterInteract);
+
+            SubscribeLocalEvent<SurgeryToolComponent, AfterInteractEvent>(OnToolAfterInteract);
+            SubscribeLocalEvent<CauteryComponent, AfterInteractEvent>(OnCauteryAfterInteract);
+        }
+
+        private void OnDrapesStartup(EntityUid uid, SurgeryDrapesComponent drapes, ComponentStartup args)
+        {
+            var ui = drapes.UserInterface;
+
+            if (ui != null)
+            {
+                ui.OnReceiveMessage += msg => OnSurgeryDrapesUIMessage(drapes, msg);
+            }
+        }
+
+        // TODO SURGERY: Add surgery for dismembered limbs
+        private void OnDrapesAfterInteract(EntityUid uid, SurgeryDrapesComponent drapes, AfterInteractEvent args)
+        {
+            var target = args.Target;
+            if (target == null)
+            {
+                return;
+            }
+
+            var user = args.User;
+            if (user.TryGetComponent(out SurgeonComponent? surgeon) &&
+                surgeon.Target != null &&
+                IsPerformingSurgeryOn(surgeon, target))
+            {
+                if (surgeon.Target.SurgeryTags.Count == 0 &&
+                    CanStopSurgery(surgeon))
+                {
+                    DoDrapesCancelPopups(drapes, surgeon, target);
+                    StopSurgery(surgeon);
+                }
+
+                args.Handled = true;
+                return;
+            }
+
+            if (!user.TryGetComponent(out ActorComponent? actor) ||
+                !target.TryGetComponent(out SharedBodyComponent? body))
```

Comment:
Same here, target's probably a bit trickier if AfterInteract doesn't already have the target entityuid.